### ISSUE TITLE
bats no longer works with programs that read stdin

### ISSUE THREAD

**Issue by natefoo** [bats-core/bats-core#197]
As of 17e2cf35ba4eb44a25f47f31f079809c0f111338 from #172 you can no longer use `run` in conjunction with programs that read from stdin, because the program will consume the remaining test files/names waiting on stdin [here](https://github.com/bats-core/bats-core/blob/8789f910812afbf6b87dd371ee5ae30592f1423f/libexec/bats-core/bats-exec-suite#L99).

A workaround is to redirect `/dev/null` to stdin, e.g. `run foo </dev/null`.

**Comment by ycdtosa** [comment]
can you please provide an example?

**Comment by natefoo** [comment]
@ycdtosa sure:

`read.sh`:
```sh
#!/bin/sh

[ ! -t 0 ] || return 0
[ -n "$DO_READ" ] || return 0

while read foo; do
    echo "$foo" >&3
done

return 0
```

`test.bats`:
```sh
#!/usr/bin/env bats

@test "test 1" {
    run ./read.sh
}

@test "test 2" {
    run ./read.sh
}

@test "test 3" {
    run ./read.sh
}
```

example run:
```console
% ./test.bats
 ✓ test 1
 ✓ test 2
 ✓ test 3

3 tests, 0 failures

% DO_READ=1 ./test.bats
 ✓ test 1

3 tests, 0 failures

% DO_READ=1 ./test.bats -t
1..3
/home/nate/test/bats/test.bats	test_test_2
/home/nate/test/bats/test.bats	test_test_3
ok 1 test 1
```

And closing stdin (updated `test.bats`, same `read.sh`):
```sh
#!/usr/bin/env bats

@test "test 1" {
    run ./read.sh </dev/null
}

@test "test 2" {
    run ./read.sh </dev/null
}

@test "test 3" {
    run ./read.sh </dev/null
}
```

example run:

```console
% DO_READ=1 ./test.bats
 ✓ test 1
 ✓ test 2
 ✓ test 3

3 tests, 0 failures

```

**Comment by cyphar** [comment]
Damn. A simple solution might be to use a different fd for the `read`.

The primary issue is that bash doesn't allow you to use pipelines while also changing bash variables (because pipelines all execute in subshells) -- and iterating over an array would be slightly nasty to get right. So we end up with the `while foo; do ; done < <()` trick.

But I'll take a look at this.

**Comment by Flamefire** [comment]
I just got bitten hard by this testing the SLURM command `srun`. I don't see why it is hard to iterate over the array. I'll give it a shot and come back with results

**Comment by Flamefire** [comment]
For a first part I made the run fail if not all tests were executed and enhanced the summary to show that: https://github.com/Flamefire/bats-core/commit/4ba5f44ca2900970a45359381d64dcca1515ba3a

**Comment by Flamefire** [comment]
And finished a fix in #227
